crm-app/src/component/customer/Profile.js

```diff
@@ -0,0 +1,179 @@
+import { connect } from 'react-redux';
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { addCustomer, updateCustomer, deleteCustomer, resetCustomer } from '../../actions';
+
+function addButton (name, handle) {
+	return (
+		<button type="button" onClick={handle}>
+			{name}
+		</button>
+	);
+}
+
+function enableElement (element) {
+	if (element.disabled) {
+		element.disabled = false;
+	}
+}
+
+function disableElement (element) {
+	if (!element.disabled) {
+		element.disabled = true;
+	}
+}
+
+class CustomerProfile extends Component {
+	constructor (props) {
+		super(props);
+		this.formRef = React.createRef();
+		this.isWritten = false;
+	}
+
+	handleSave = (e) => {
+		const { displayedIndex, onAdd, onUpdate } = this.props;
+		const form = e.target.form;
+		const name = form.name.value;
+		const mail = form.mail.value;
+		const phone = form.phone.value;
+
+		if (name === '' || mail === '' || phone === '') {
+			return;
+		}
+
+		// check if it is new customer or modified customer
+		if (this.isWritten) {
+			onUpdate(displayedIndex, name, phone);
+		} else {
+			if (displayedIndex < 0) {
+				onAdd(name, mail, phone);
+			} else {
+				alert('Error: This email alreay exists.');
+			}
+		}
+
+		e.preventDefault();
+	}
+
+	handleReset = (e) => {
+		this.clearForm(e.target.form);
+		this.props.onReset();
+	}
+
+	handleDelete = (e) => {
+		const { displayedIndex, onDelete } = this.props;
+		onDelete(displayedIndex);
+	}
+
+	clearForm = (target) => {
+		enableElement(target.mail);
+		target.name.value = '';
+		target.mail.value = '';
+		target.phone.value = '';
+	}
+
+	componentDidMount () {
+		const form = this.formRef.current;
+		form.submit.addEventListener('click', this.handleSave);
+		form.reset.addEventListener('click', this.handleReset);
+	}
+
+	componentWillUpdate (nextProps, nextState) {
+		const {customers, displayedIndex } = nextProps;
+		const form = this.formRef.current;
+
+		if (displayedIndex < 0) {
+			this.clearForm(form);
+			this.isWritten = false;
+			this.deleteButton = null;
+		} else {
+			form.name.value = customers[displayedIndex].name;
+			form.mail.value = customers[displayedIndex].mail;
+			form.phone.value = customers[displayedIndex].phone;
+			this.isWritten = true;
+			disableElement(form.mail);
+			this.deleteButton = addButton('delete', this.handleDelete);
```

Comment:
In IMHO, `deleteButton` would be managed as a state because it is related with rendering.

Reply:
I'm trying to refactor them when I implement this application with Enact framework.